### PR TITLE
[SPARK-20067] [SQL] Unify and Clean Up Desc Commands Using Catalog Interface

### PR DESCRIPTION
**Pull request by gatorsmile**
### What changes were proposed in this pull request?

This PR is to unify and clean up the outputs of `DESC EXTENDED/FORMATTED` and `SHOW TABLE EXTENDED` by moving the logics into the Catalog interface. The output formats are improved. We also add the missing attributes. It impacts the DDL commands like `SHOW TABLE EXTENDED`, `DESC EXTENDED` and `DESC FORMATTED`.

In addition, by following what we did in Dataset API `printSchema`, we can use `treeString` to show the schema in the more readable way. 

Below is the current way:
```
Schema: STRUCT<`a`: STRING (nullable = true), `b`: INT (nullable = true), `c`: STRING (nullable = true), `d`: STRING (nullable = true)>
```
After the change, it should look like
```
Schema: root
 |-- a: string (nullable = true)
 |-- b: integer (nullable = true)
 |-- c: string (nullable = true)
 |-- d: string (nullable = true)
```

### How was this patch tested?
`describe.sql` and `show-tables.sql`

### CODE REVIEW

**Comment by gatorsmile** [comment]
cc @cloud-fan

File: sql/core/src/test/resources/sql-tests/results/describe.sql.out [hunk 1]
```diff
@@ -97,16 +97,21 @@ DESC EXTENDED t
 struct<col_name:string,data_type:string,comment:string>
 -- !query 6 output
 # Detailed Table Information	CatalogTable(
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/describe.sql.out]: do we need to wrap the information with `CatalogTable(...)`?

Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/results/describe.sql.out]: Yes. Let me remove it and move it to the next line

Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/results/describe.sql.out]: The output looks weird if we sort the Command, so file a PR: https://github.com/apache/spark/pull/17414

File: sql/core/src/test/resources/sql-tests/results/describe.sql.out [hunk 2]
```diff
@@ -68,67 +68,74 @@ DESC FORMATTED t
...
+Provider: parquet
+Partition Columns: [`c`, `d`]
+Comment: table_comment
+Storage(Location: sql/core/spark-warehouse/t)
+Partition Provider: Catalog
+Schema: root
+ |-- a: string (nullable = true)
+ |-- b: integer (nullable = true)
+ |-- c: string (nullable = true)
+ |-- d: string (nullable = true)
```
Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/results/describe.sql.out]: @cloud-fan Do we still need `Schema`, if the above info already has it? The only missing part is the nullability

Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/describe.sql.out]: yea we do not need, but I think `CatalogTable.toString` should still have the schema. Actually why do we have `DESC EXTENDED`? I think `DESC FORMATTED` is good enough

**Comment by gatorsmile** [comment]
retest this please

File: sql/core/src/test/resources/sql-tests/results/describe.sql.out (hunk 2 above)
Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/results/describe.sql.out]: Below is what `Programming Hive` said:
> Replacing EXTENDED with FORMATTED provides more readable but also more verbose output.

For most end-users, they should use `DESC FORMATTED `.

Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/results/describe.sql.out]: Maybe we can keep the output of `DESC EXTENDED` as it is. Then, we need to revisit `DESC FORMATTED` to ensure it includes all the user-facing metadata info.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala [hunk 3]
```diff
@@ -273,28 +278,32 @@ case class CatalogTable(
           if (bucketColumnNames.nonEmpty) s"Bucket Columns: $bucketColumnsString" else "",
           if (sortColumnNames.nonEmpty) s"Sort Columns: $sortColumnsString" else ""
         )
-
       case _ => Nil
     }
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: We should de-duplicate the codes.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala [hunk 4]
```diff
@@ -21,6 +21,7 @@ import java.net.URI
 import java.util.Date
 
 import com.google.common.base.Objects
+import scala.collection.mutable
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: Usually we put the imports of scala library above third-party.

**Comment by viirya** [comment]
Please update the PR description too since you update the title.

**Comment by cloud-fan** [comment]
after this PR, what's the difference between `DESC EXTENDED` and `DESC FORMATTED`?

**Comment by gatorsmile** [comment]
After refactoring, I think `DESC FORMATTED` might be more useful for users.

`DESC FORMATTED` outputs the metadata in a more formatted way. Users can query it using the key value (stored in `col_name`) to get the result. They do not need to parse the string to extract the data value, unlike `DESC EXTENDED`.

Below is the output of `DESC FORMATTED `

```
+----------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------+-------+
|col_name                    |data_type                                                                                                                                         |comment|
+----------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------+-------+
|key                         |int                                                                                                                                               |test   |
|value                       |string                                                                                                                                            |null   |
|ds                          |string                                                                                                                                            |null   |
|hr                          |string                                                                                                                                            |null   |
|# Partition Information     |                                                                                                                                                  |       |
|# col_name                  |data_type                                                                                                                                         |comment|
|ds                          |string                                                                                                                                            |null   |
|hr                          |string                                                                                                                                            |null   |
|                            |                                                                                                                                                  |       |
|# Detailed Table Information|                                                                                                                                                  |       |
|Table                       |`default`.`tab1`                                                                                                                                  |       |
|Owner                       |xiao                                                                                                                                              |       |
|Created                     |Mon Mar 27 21:25:07 PDT 2017                                                                                                                      |       |
|Last Access                 |Wed Dec 31 16:00:00 PST 1969                                                                                                                      |       |
|Type                        |EXTERNAL                                                                                                                                          |       |
|Provider                    |hive                                                                                                                                              |       |
|Comment                     |Apache Spark                                                                                                                                      |       |
|Properties                  |[serialization.format=1]                                                                                                                          |       |
|Location                    |file:/private/var/folders/6r/15tqm8hn3ldb3rmbfqm1gf4c0000gn/T/spark-8d941400-8f59-4080-8094-611fa517f11e                                          |       |
|Serde Library               |org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe                                                                                                |       |
|InputFormat                 |org.apache.hadoop.mapred.TextInputFormat                                                                                                          |       |
|OutputFormat                |org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat                                                                                        |       |
|Partition Provider          |Catalog                                                                                                                                           |       |
|Partition Columns           |[`ds`, `hr`]                                                                                                                                      |       |
|Schema                      |root
 |-- key: integer (nullable = true)
 |-- value: string (nullable = true)
 |-- ds: string (nullable = true)
 |-- hr: string (nullable = true)
|       |
+----------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------+-------+
```

Below is the output of `DESC EXTENDED`

```
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------+-------+
|col_name                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               |data_type|comment|
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------+-------+
|key                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |int      |test   |
|value                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  |string   |null   |
|ds                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |string   |null   |
|hr                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |string   |null   |
|# Partition Information                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                |         |       |
|# col_name                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             |data_type|comment|
|ds                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |string   |null   |
|hr                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |string   |null   |
|                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       |         |       |
|# Detailed Table Information                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |         |       |
|Table: `default`.`tab1`
Owner: xiao
Created: Mon Mar 27 21:25:07 PDT 2017
Last Access: Wed Dec 31 16:00:00 PST 1969
Type: EXTERNAL
Provider: hive
Comment: Apache Spark
Properties: [serialization.format=1]
Location: file:/private/var/folders/6r/15tqm8hn3ldb3rmbfqm1gf4c0000gn/T/spark-8d941400-8f59-4080-8094-611fa517f11e
Serde Library: org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe
InputFormat: org.apache.hadoop.mapred.TextInputFormat
OutputFormat: org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat
Partition Provider: Catalog
Partition Columns: [`ds`, `hr`]
Schema: root
 |-- key: integer (nullable = true)
 |-- value: string (nullable = true)
 |-- ds: string (nullable = true)
 |-- hr: string (nullable = true)
|         |       |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------+-------+


```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala (hunk 4 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: If we change the order, can scalastyle pass?

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: Usually, the import order is:

    import java.util.Date

    import scala.collection.mutable

    import com.google.common.base.Objects

`Dataset.scala` is an example.

**Comment by cloud-fan** [comment]
seems `DESC EXTENDED` is totally useless... How about we deprecate `DESC EXTENDED` at parser and implement it with `DESC FORMATTED`?

**Comment by gatorsmile** [comment]
Yes. We do not need to keep the difference between `DESC EXTENDED` and `DESC FORMATTED`

**Comment by gatorsmile** [comment]
Hive DESC TABLE has different formats. We hit many error due to it. I think we can disable it when comparing the results between Hive and Spark SQL.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala [hunk 5]
```diff
@@ -500,7 +500,6 @@ case class TruncateTableCommand(
 case class DescribeTableCommand(
     table: TableIdentifier,
     partitionSpec: TablePartitionSpec,
-    isExtended: Boolean,
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala]: seems `DESC EXTENDED` is better, normal `DESC tbl` is also formated

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala [hunk 6]
```diff
@@ -91,15 +98,27 @@ case class CatalogTablePartition(
...
+  }
 
-    output.filter(_.nonEmpty).mkString("CatalogPartition(\n\t", "\n\t", ")")
+  override def toString: String = {
+    toLinkedHashMap.map { case ((key, value)) =>
+      if (value.isEmpty) key else s"$key: $value"
+    }.mkString("CatalogPartition(\n\t", "\n\t", ")")
+  }
+
+  def simpleString: String = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: when shall we call `simpleString`?

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: `Show Table` command will call it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala (hunk 5 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala]: Sure, let me make the change.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala [hunk 7]
```diff
@@ -214,6 +215,7 @@ class SQLQueryTestSuite extends QueryTest with SharedSQLContext {
     // Returns true if the plan is supposed to be sorted.
     def isSorted(plan: LogicalPlan): Boolean = plan match {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala]: maybe call it `needSort`?

File: sql/core/src/test/resources/sql-tests/results/describe.sql.out [hunk 8]
```diff
@@ -1,205 +1,248 @@
...
+d                   	string              	                    
+# Partition Information	                    	                    
+# col_name          	data_type           	comment             
 c                   	string              	                    
 d                   	string              	                    
-d                   	string
+                    	                    	                    
+# Detailed Table Information	                    	                    
+Table               	`default`.`t`       	                    
+Created[removed by SQLQueryTestSuite]
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/describe.sql.out]: this looks weird, how about `Created [not included in comparison]`

File: sql/core/src/test/resources/sql-tests/results/describe.sql.out [hunk 9]
```diff
@@ -1,205 +1,248 @@
...
+Type                	MANAGED             	                    
+Provider            	parquet             	                    
+Num Buckets         	2                   	                    
+Bucket Columns      	[`a`]               	                    
+Sort Columns        	[`b`]               	                    
+Comment             	table_comment       	                    
+Location[removed by SQLQueryTestSuite]sql/core/spark-warehouse/t	                    
+Partition Provider  	Catalog             	                    
+Partition Columns   	[`c`, `d`]          	                    
+Schema              	root
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/describe.sql.out]: this information is duplicated

File: sql/core/src/test/resources/sql-tests/results/describe.sql.out [hunk 10]
```diff
@@ -1,205 +1,248 @@
...
 c                   	string              	                    
+d                   	string              	                    
+# Partition Information	                    	                    
+# col_name          	data_type           	comment             
 c                   	string              	                    
 d                   	string              	                    
-d                   	string
+                    	                    	                    
+# Detailed Table Information	                    	                    
+Table               	`default`.`t`
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/describe.sql.out]: shall we also include the database?

Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/describe.sql.out]: shall we separate it into 2 lines, `Database` and `Table`?

File: sql/core/src/test/resources/sql-tests/results/describe.sql.out [hunk 11]
```diff
@@ -1,205 +1,248 @@
...
 -- !query 8 schema
 struct<col_name:string,data_type:string,comment:string>
 -- !query 8 output
-# Partition Information	                    	                    
-# col_name          	data_type           	comment             
-Detailed Partition Information CatalogPartition(
-	Partition Values: [c=Us, d=1]
-	Storage(Location: sql/core/spark-warehouse/t/c=Us/d=1)
-	Partition Parameters:{})	                    	                    
 a                   	string
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/describe.sql.out]: we should add a header here `# col_name          	data_type           	comment`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala [hunk 12]
```diff
@@ -637,21 +570,7 @@ case class DescribeTableCommand(
...
-  }
-
-  private def describeExtendedDetailedPartitionInfo(
-      tableIdentifier: TableIdentifier,
-      table: CatalogTable,
-      partition: CatalogTablePartition,
-      buffer: ArrayBuffer[Row]): Unit = {
-    append(buffer, "", "", "")
-    append(buffer, "Detailed Partition Information " + partition.toString, "", "")
+    if (isExtended) describeFormattedDetailedPartitionInfo(table, metadata, partition, result)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala]: not related to this PR, but it looks weird that `DESC tbl` and `DESC tbl PARTITION (xxx)` has the same result.

File: sql/core/src/test/resources/sql-tests/results/describe.sql.out [hunk 13]
```diff
@@ -1,205 +1,248 @@
...
+d                   	string              	                    
+# Partition Information	                    	                    
+# col_name          	data_type           	comment             
 c                   	string              	                    
 d                   	string              	                    
-d                   	string
+                    	                    	                    
+# Detailed Partition Information	                    	                    
+Database            	default             	                    
+Table               	t
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/describe.sql.out]: not related to this PR, but why do we put the `Database` and `Table` in `# Detailed Partition Information`? Did we follow hive?

Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/results/describe.sql.out]: Yes. 
```
hive> DESC EXTENDED page_view PARTITION(dt='part1', country='part2');
OK
viewtime            	int                 	                    
userid              	bigint              	                    
page_url            	string              	                    
referrer_url        	string              	                    
ip                  	string              	IP Address of the User
dt                  	string              	                    
country             	string              	                    
	 	 
# Partition Information	 	 
# col_name            	data_type           	comment             
	 	 
dt                  	string              	                    
country             	string              	                    
	 	 
Detailed Partition Information	Partition(values:[part1, part2], dbName:default, tableName:page_view, createTime:1491066325, lastAccessTime:0, sd:StorageDescriptor(cols:[FieldSchema(name:viewtime, type:int, comment:null), FieldSchema(name:userid, type:bigint, comment:null), FieldSchema(name:page_url, type:string, comment:null), FieldSchema(name:referrer_url, type:string, comment:null), FieldSchema(name:ip, type:string, comment:IP Address of the User), FieldSchema(name:dt, type:string, comment:null), FieldSchema(name:country, type:string, comment:null)], location:file:/user/hive/warehouse/page_view/dt=part1/country=part2, inputFormat:org.apache.hadoop.mapred.SequenceFileInputFormat, outputFormat:org.apache.hadoop.hive.ql.io.HiveSequenceFileOutputFormat, compressed:false, numBuckets:-1, serdeInfo:SerDeInfo(name:null, serializationLib:org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe, parameters:{serialization.format=1}), bucketCols:[], sortCols:[], parameters:{}, skewedInfo:SkewedInfo(skewedColNames:[], skewedColValues:[], skewedColValueLocationMaps:{}), storedAsSubDirectories:false), parameters:{totalSize=0, numRows=0, rawDataSize=0, COLUMN_STATS_ACCURATE={"BASIC_STATS":"true"}, numFiles=0, transient_lastDdlTime=1491066325})	
```

File: sql/core/src/test/resources/sql-tests/results/describe.sql.out (hunk 9 above)
Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/results/describe.sql.out]: Let me remove both

File: sql/core/src/test/resources/sql-tests/results/describe.sql.out (hunk 10 above)
Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/results/describe.sql.out]: Sure

File: sql/core/src/test/resources/sql-tests/results/describe.sql.out (hunk 8 above)
Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/results/describe.sql.out]: Sure

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala (hunk 12 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala]: This function `describeDetailedPartitionInfo ` will only be called for the DDL command
```SQL
  DESCRIBE [EXTENDED|FORMATTED] table_name PARTITION (partitionVal*)
```

File: sql/core/src/test/resources/sql-tests/results/describe.sql.out (hunk 11 above)
Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/results/describe.sql.out]: Done.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/DDLTestSuite.scala [hunk 14]
```diff
@@ -1,123 +0,0 @@
...
-      sqlContext: SQLContext,
-      parameters: Map[String, String]): BaseRelation = {
-    SimpleDDLScan(
-      parameters("from").toInt,
-      parameters("TO").toInt,
-      parameters("Table"))(sqlContext.sparkSession)
-  }
-}
-
-case class SimpleDDLScan(
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/DDLTestSuite.scala]: These two classes are moved to `DataSourceTest.scala`

File: sql/core/src/test/resources/sql-tests/results/describe.sql.out [hunk 15]
```diff
@@ -1,205 +1,259 @@
...
+Bucket Columns      	[`a`]               	                    
+Sort Columns        	[`b`]               	                    
+Comment             	table_comment       	                    
+Location [not included in comparison]sql/core/spark-warehouse/t	                    
+Partition Provider  	Catalog
 
 
--- !query 7
+-- !query 10
 DESC t PARTITION (c='Us', d=1)
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/describe.sql.out]: This has the same result with `DESC t`, is it same with Hive?

File: sql/core/src/test/resources/sql-tests/results/describe.sql.out [hunk 16]
```diff
@@ -1,205 +1,259 @@
...
 d                   	string              	                    
-d                   	string
+                    	                    	                    
+# Detailed Partition Information	                    	                    
+Database            	default             	                    
+Table               	t                   	                    
+Partition Values    	[c=Us, d=1]         	                    
+Location [not included in comparison]sql/core/spark-warehouse/t/c=Us/d=1	                    
+                    	                    	                    
+# Table Storage Information
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/describe.sql.out]: why we have a new `# Table Storage Information` section for `DESC EXTENDED t PARTITION (xxx)`?

File: sql/core/src/test/resources/sql-tests/results/describe.sql.out (hunk 15 above)
Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/results/describe.sql.out]: Yes. 
```
hive> DESC page_view PARTITION(dt='part1', country='part2');
OK
viewtime            	int                 	                    
userid              	bigint              	                    
page_url            	string              	                    
referrer_url        	string              	                    
ip                  	string              	IP Address of the User
dt                  	string              	                    
country             	string              	                    
	 	 
# Partition Information	 	 
# col_name            	data_type           	comment             
	 	 
dt                  	string              	                    
country             	string              	                    
Time taken: 0.12 seconds, Fetched: 13 row(s)
hive> DESC page_view;
OK
viewtime            	int                 	                    
userid              	bigint              	                    
page_url            	string              	                    
referrer_url        	string              	                    
ip                  	string              	IP Address of the User
dt                  	string              	                    
country             	string              	                    
	 	 
# Partition Information	 	 
# col_name            	data_type           	comment             
	 	 
dt                  	string              	                    
country             	string              	                    
Time taken: 0.022 seconds, Fetched: 13 row(s)
```

File: sql/core/src/test/resources/sql-tests/results/describe.sql.out (hunk 16 above)
Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/results/describe.sql.out]: Hive does it. See the following output.
```
hive> DESC FORMATTED page_view PARTITION(dt='part1', country='part2');
OK
# col_name            	data_type           	comment             
	 	 
viewtime            	int                 	                    
userid              	bigint              	                    
page_url            	string              	                    
referrer_url        	string              	                    
ip                  	string              	IP Address of the User
	 	 
# Partition Information	 	 
# col_name            	data_type           	comment             
	 	 
dt                  	string              	                    
country             	string              	                    
	 	 
# Detailed Partition Information	 	 
Partition Value:    	[part1, part2]      	 
Database:           	default             	 
Table:              	page_view           	 
CreateTime:         	Sat Apr 01 17:05:25 UTC 2017	 
LastAccessTime:     	UNKNOWN             	 
Location:           	file:/user/hive/warehouse/page_view/dt=part1/country=part2	 
Partition Parameters:	 	 
	COLUMN_STATS_ACCURATE	{\"BASIC_STATS\":\"true\"}
	numFiles            	0                   
	numRows             	0                   
	rawDataSize         	0                   
	totalSize           	0                   
	transient_lastDdlTime	1491066325          
	 	 
# Storage Information	 	 
SerDe Library:      	org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe	 
InputFormat:        	org.apache.hadoop.mapred.SequenceFileInputFormat	 
OutputFormat:       	org.apache.hadoop.hive.ql.io.HiveSequenceFileOutputFormat	 
Compressed:         	No                  	 
Num Buckets:        	-1                  	 
Bucket Columns:     	[]                  	 
Sort Columns:       	[]                  	 
Storage Desc Params:	 	 
	serialization.format	1                   
Time taken: 0.124 seconds, Fetched: 39 row(s)

hive> DESC EXTENDED page_view PARTITION(dt='part1', country='part2');
OK
viewtime            	int                 	                    
userid              	bigint              	                    
page_url            	string              	                    
referrer_url        	string              	                    
ip                  	string              	IP Address of the User
dt                  	string              	                    
country             	string              	                    
	 	 
# Partition Information	 	 
# col_name            	data_type           	comment             
	 	 
dt                  	string              	                    
country             	string              	                    
	 	 
Detailed Partition Information	Partition(values:[part1, part2], dbName:default, tableName:page_view, createTime:1491066325, lastAccessTime:0, sd:StorageDescriptor(cols:[FieldSchema(name:viewtime, type:int, comment:null), FieldSchema(name:userid, type:bigint, comment:null), FieldSchema(name:page_url, type:string, comment:null), FieldSchema(name:referrer_url, type:string, comment:null), FieldSchema(name:ip, type:string, comment:IP Address of the User), FieldSchema(name:dt, type:string, comment:null), FieldSchema(name:country, type:string, comment:null)], location:file:/user/hive/warehouse/page_view/dt=part1/country=part2, inputFormat:org.apache.hadoop.mapred.SequenceFileInputFormat, outputFormat:org.apache.hadoop.hive.ql.io.HiveSequenceFileOutputFormat, compressed:false, numBuckets:-1, serdeInfo:SerDeInfo(name:null, serializationLib:org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe, parameters:{serialization.format=1}), bucketCols:[], sortCols:[], parameters:{}, skewedInfo:SkewedInfo(skewedColNames:[], skewedColValues:[], skewedColValueLocationMaps:{}), storedAsSubDirectories:false), parameters:{totalSize=0, numRows=0, rawDataSize=0, COLUMN_STATS_ACCURATE={"BASIC_STATS":"true"}, numFiles=0, transient_lastDdlTime=1491066325})	
Time taken: 0.116 seconds, Fetched: 15 row(s)
```

Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/describe.sql.out]: hive use `# Storage Information`, which seems to be the storage of the partition, not the table

Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/results/describe.sql.out]: Yes. Just checked the [source code of Hive](https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/metadata/formatting/MetaDataFormatUtils.java#L348-L352). This section in Hive is the storage for Partition, but the info was originally from the table storage, as shown in [the codes](https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/metadata/Partition.java#L94-L96). We did the same thing. 

The only difference is our `BucketSpec` is not part of `CatalogStorageFormat`. Should we follow Hive to move it into `CatalogStorageFormat`? Or do it as a follow-up? After the move, we can easily output the exactly the same things as Hive.

Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/describe.sql.out]: Shall we rename this section to `Storage Information`? We may have per-partition storage in the future, although for now this is always table storage.

Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/results/describe.sql.out]: Sure, let me make the change.

**Comment by cloud-fan** [comment]
LGTM

**Comment by gatorsmile** [comment]
Thanks! Merging to master.

**Comment by dongjoon-hyun** [comment]
Hi, All .
This seems to be incompatible changes even in `DESC t` from 2.2.0.
Is it better to be compatible on `DESC t` which has no `EXTENDED|FORMATTED`?

**Comment by gatorsmile** [comment]
Add the following line when issuing `DESC t` on partitioned tables?
` # col_name          	data_type           	comment`

**Comment by dongjoon-hyun** [comment]
Yep. That one.

**Comment by dongjoon-hyun** [comment]
It's reported at https://issues.apache.org/jira/browse/SPARK-20954 .
If the current one is better, could you give some advice on that? :)

**Comment by dongjoon-hyun** [comment]
What I mean is to removing `# col_name data_type comment` in case of `DESC t`.

**Comment by gatorsmile** [comment]
Sure, we can remove it. Feel free to submit a PR.

**Comment by dongjoon-hyun** [comment]
Thank you for advice. I'll make a PR soon.

**Comment by dongjoon-hyun** [comment]
Hi, All.
I made a PR, https://github.com/apache/spark/pull/18203.

**Pull request closed without merging**